optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,241 @@
+                .join(Exam.class,
+                        equal(TopicConflict::getLeftTopic, Exam::getTopic),
+                        filtering((topicConflict, leftExam) -> leftExam.getPeriod() != null))
+                .ifExists(Exam.class,
+                        equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex),
+                        // Find exams in a day, but not being held right after each other. That case is handled in the twoExamsInARow constraint.
+                        filtering((topicConflict, leftExam,
+                                rightExam) -> getPeriodIndexDifferenceBetweenExams(leftExam, rightExam) > 1))
+                .join(ExaminationConstraintConfiguration.class)
+                .penalize("Two exams in a day",
+                        HardSoftScore.ONE_SOFT,
+                        (topicConflict, leftExam, config) -> topicConflict.getStudentSize() * config.getTwoInADayPenalty());
+    }
+
+    protected Constraint periodSpread(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ExaminationConstraintConfiguration.class)
+                .join(TopicConflict.class)
+                .join(Exam.class,
+                        equal((config, topicConflict) -> topicConflict.getLeftTopic(), Exam::getTopic),
+                        filtering((config, topicConflict, leftExam) -> leftExam.getPeriod() != null))
+                .ifExists(Exam.class,
+                        equal((config, topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        filtering((config, topicConflict, leftExam, rightExam) -> rightExam.getPeriod() != null),
+                        filtering((config, topicConflict, leftExam,
+                                rightExam) -> getPeriodIndexDifferenceBetweenExams(leftExam,
+                                        rightExam) < (config.getPeriodSpreadLength() + 1)))
+                .penalize("Period spread",
+                        HardSoftScore.ONE_SOFT,
+                        (config, topicConflict, leftExam) -> topicConflict.getStudentSize());
+    }
+
+    protected Constraint mixedDurations(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
+                .ifNotExistsOther(Exam.class,
+                        equal(Exam::getPeriod, Exam::getPeriod),
+                        equal(Exam::getRoom, Exam::getRoom),
+                        lessThan(Exam::getId, Exam::getId))
+                .join(Exam.class,
+                        equal(Exam::getPeriod),
+                        equal(Exam::getRoom),
+                        greaterThan(Exam::getId),
+                        filtering((leftExam, rightExam) -> rightExam.getTopicDuration() != leftExam.getTopicDuration()))
```

Comment:
Also, I wonder why this entire construct isn't simply `.fromUniquePair(Exam.class, ...)`